sys/dev/usb/uvc/uvc_v4l2.c

```diff
@@ -0,0 +1,784 @@
+
+	memset(arg, 0, sizeof(*arg));
+	arg->type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+
+	arg->parm.capture.capability = V4L2_CAP_TIMEPERFRAME;
+	arg->parm.capture.capturemode = 0;
+	arg->parm.capture.timeperframe.numerator = numerator;
+	arg->parm.capture.timeperframe.denominator = denominator;
+	arg->parm.capture.extendedmode = 0;
+	arg->parm.capture.readbuffers = 0;
+
+	return 0;
+}
+
+//only stub function
+static int
+uvc_v4l2_enum_input(struct uvc_drv_ctrl *ctrl, struct v4l2_input *input)
+{
+	uint32_t index = input->index;
+	struct uvc_drv_entity *it;
+
+	/*WARNING: to_be_implement here, hard code no SELECTOR */
+	if (index)
+		return EINVAL;
+
+	it = STAILQ_FIRST(&ctrl->entities);
+	if (!it)
+		return EINVAL;
+
+	memset(input, 0, sizeof(*input));
+	input->index = index;
+	strlcpy(input->name, it->name, sizeof(input->name));
+	input->type = V4L2_INPUT_TYPE_CAMERA;
+
+	return 0;
+}
+
+static void
+uvc_v4l2_dtor(void *data)
+{
+	struct uvc_v4l2_cdev_priv *priv = data;
+	struct uvc_drv_video *v;
+
+	v = priv->v;
+	if (priv->work_pri == UVC_V4L2_PRI_ACTIVE) {
+		atomic_store_64(&v->pri, 0);
+	}
+
+	atomic_subtract_64(&v->users, 1);
+
+	free(data, M_UVC);
+
+	DPRINTF("%s\n", __func__);
+}
+
+static int
+uvc_v4l2_queryctrl(struct uvc_drv_video *v, struct v4l2_queryctrl *qc)
+{
+	int ret = EINVAL;
+
+	if (qc == NULL)
+		return ret;
+
+	ret = uvc_query_v4l2_ctrl(v, qc);
+	return ret;
+}
+
+static int
+uvc_v4l2_querymenu(struct uvc_drv_video *v, struct v4l2_querymenu *qm)
+{
+	int ret = EINVAL;
+
+	if (qm == NULL)
+		return ret;
+
+	ret = uvc_query_v4l2_menu(v, qm);
+	return ret;
+}
+static int
+uvc_v4l2_open(struct cdev *dev, int flags, int fmt, struct thread *td)
+{
+	struct uvc_v4l2_cdev_priv *priv;
+	struct uvc_drv_video *v = dev->si_drv1;
+
+	if (v4l2_not_allowed != 0) {
+		return ENXIO;			/* failure */
+	}
+
+	atomic_add_64(&v->users, 1);
+
+	DPRINTF("===v4l2 open %s-%s %d-%d mem:%lu pri:%lu===\n",
+		curthread->td_proc->p_comm, curthread->td_proc->p_pptr->p_comm,
+		curthread->td_proc->p_pid, curthread->td_proc->p_pptr->p_pid,
+		v->users, v->pri);
+
+	DPRINTF("====%s====%d %d %s\n", __func__, curthread->td_proc->p_pid,
+		curthread->td_proc->p_pptr->p_pid,
+		curthread->td_proc->p_pptr->p_comm);
+	DPRINTF("enable:%lu\n", v->enable);
+
+	priv = (struct uvc_v4l2_cdev_priv *)
+		malloc(sizeof(*priv), M_UVC, M_ZERO | M_WAITOK);
+	if (!priv) {
+		DPRINTF("%s %d------>Error.\n", __func__, __LINE__);
+		atomic_subtract_64(&v->users, 1);
+		return ENOMEM;
+	}
+
+	priv->work_mode = UVC_V4L2_MODE_READ;
+	priv->work_pri = UVC_V4L2_PRI_PASSIVE;
+	priv->v = v;
+	priv->num = v->users;
+	(void)devfs_set_cdevpriv(priv, uvc_v4l2_dtor);
+	return 0;
+}
+
+static int
+uvc_v4l2_close(struct cdev *dev, int flags, int fmt, struct thread *td)
+{
+	struct uvc_drv_video *v = dev->si_drv1;
+	struct uvc_v4l2_cdev_priv *priv;
+	int ret;
+
+	ret = devfs_get_cdevpriv((void **)&priv);
+	if (ret != 0) {
+		DPRINTF("error===================================\n");
```

Comment:
Sneaked debug code? Print more details, e.g., `ret` value is recommended, or remove it.